File: 백준/Silver/1926. 그림/그림.py
import sys
from collections import deque
input = sys.stdin.readline


n, m = map(int, input().split())
paper = [list(map(int, input().split())) for _ in range(n)]

dxs, dys = [0, 1, 0, -1], [1, 0, -1, 0]

def bfs(x, y):
    area = 1
    queue = deque([(x, y)])
    paper[x][y] = 0 # visited

    while queue:
        cx, cy = queue.popleft()

        for dx, dy in zip(dxs, dys):
            nx, ny = dx + cx, dy + cy

            if 0 <= nx < n and 0 <= ny < m and paper[nx][ny] == 1:
                area += 1
                paper[nx][ny] = 0
                queue.append((nx, ny))

    return area


count = 0 # 그림의 개수
max_area = 0

for i in range(n):
    for j in range(m):
        if paper[i][j] == 1:
            count += 1
            max_area = max(max_area, bfs(i, j))

print(count)
print(max_area)

File: 백준/Silver/1926. 그림/그림.java
import java.util.*;
public class Main {

    static int[] dxs = {-1, 0, 1, 0};
    static int[] dys = {0, 1, 0, -1};
    static int[][] paper;
    static int n, m;

    public static int bfs(int x, int y) {
        int area = 1;
        Queue<int[]> queue = new LinkedList<>();
        queue.add(new int[]{x, y});
        paper[x][y] = 0; // 방문 처리

        while (!queue.isEmpty()) {
            int[] current = queue.poll();
            int cx = current[0];
            int cy = current[1];

            for (int i = 0; i < 4; i++) {
                int nx = cx + dxs[i];
                int ny = cy + dys[i];

                if (0 <= nx && nx < n && 0 <= ny && ny < m && paper[nx][ny] == 1) {
                    area++;
                    paper[nx][ny] = 0;
                    queue.add(new int[]{nx, ny});
                }
            }
        }
        return area;
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        n = sc.nextInt();
        m = sc.nextInt();

        paper = new int[n][m];
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                paper[i][j] = sc.nextInt();
            }
        }

        int count = 0;
        int maxArea = 0;

        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if (paper[i][j] == 1) {
                    count++;
                    maxArea = Math.max(maxArea, bfs(i, j));
                }
            }
        }

        System.out.println(count);
        System.out.println(maxArea);

        sc.close();
    }
}
